arch/risc-v/src/common/riscv_macros.S

```diff
@@ -227,8 +222,15 @@
   REGLOAD      t0, REG_INT_CTX(\out)
   li           t1, MSTATUS_FS
   and          t2, t0, t1
-  li           t1, MSTATUS_FS_INIT
-  ble          t2, t1, 1f
+  li           t1, MSTATUS_FS_DIRTY
+  bne          t2, t1, 1f
+
+  /* Reset FS bit to MSTATUS_FS_CLEAN */
+  li           t1, MSTATUS_FS_CLEAN
```

Comment:
but this change makes the thread fail to restore the FPU context in the following case:

1. Thread 1 use FPU and then switch out(FPU save with dirty flag)
2. Thread 1 switch in some time later(FPU restore with clean flag)
3. Thread 1 doesn't use FPU in this time and switch out(skip FPU saving)
4. Thread 2 swith in use FPU and thread 1 FPU context is destroyed
5. Thread 1 switch in again, but can't restore the FPU context

So, the root cause is that https://github.com/apache/nuttx/pull/5731 move the storage of context from tcb to stack, and then we can't save FPU one time but restore FPU multiple time because the stack(FPU saved here) may be modified after the thread resume the execution.

To fix this problem, we need save/restore FPU context every time if MSTATUS_FS doesn't equal MSTATUS_FS_INIT.

Reply:
Would it be possible to save FPU registers into the tcb again and save/restore FPU state only when a switch out/in happens?

Reply:
> 4. Thread 2 switch in use FPU and thread 1 FPU context(hardware register) is destroyed
> 5. Thread 1 switch in again, but can't restore the FPU context

During "5. Thread 1 switch in", the FS bit is in a clean state. Even if the FPU registers have been corrupted at this point, would it have any unknown impact on the possible execution of FPU instructions by Thread 1?

Reply:
> Would it be possible to save FPU registers into the tcb again and save/restore FPU state only when a switch out/in happens?

Yes, it could be, but it waste the memory for thread which never use FPU.

Reply:
Does it not do it now anyway? Not by increasing the size of the tcb, but every program needs a bigger stack ? Because on trap/exception entry, room for FPU regs is reserved whether FPU is used or not.

Reply:
The implementation isn't optimized as arm version, you can see:
https://github.com/apache/nuttx/blob/master/arch/arm/src/armv7-m/arm_exception.S#L181C18-L182
arm version doesn't reserve the space for FPU in this case.
@no1wudi could you apply the same optimization to riscv?

Reply:
OK

Reply:
> but this change makes the thread fail to restore the FPU context in the following case:
> 
> 1. Thread 1 use FPU and then switch out(FPU save with dirty flag)
> 2. Thread 1 switch in some time later(FPU restore with clean flag)
> 3. Thread 1 doesn't use FPU in this time and switch out(skip FPU saving)
> 4. Thread 2 switch in use FPU and thread 1 FPU context(hardware register) is destroyed
> 5. Thread 1 switch in again, but can't restore the FPU context
> 
> So, the root cause is that #5731 move the storage of context from tcb to stack, and then we can't save FPU once but restore FPU multiple time because the stack(FPU saved here) may be modified after the thread resume the execution.
> 
> To fix this problem, we need save/restore FPU context every time if MSTATUS_FS doesn't equal MSTATUS_FS_INIT.

Can we consider not cleaning the FS bit after restoring FPU?

Reply:
do you mean keep dirtly?

Reply:
> do you mean keep dirtly?

Yes, space has always been reserved for FPU in the stack.

Reply:
we need save/restore FPU context after the thread run any fpu instruction. Yes, the code doesn't need clear the flag anymore.

Reply:
I agree that with the current implementation FPU needs to be saved & restored always if the thread ever uses the FPU, even once. Otherwise the operation is never atomic and the restore occurs from the wrong stack frame. The current "lazy fpu" implementation is just plain wrong and needs to be removed completely in order for this to work.

However, I would still like to discuss doing the lazy FPU properly, by adding a separate FPU save area into the TCB. RISC-V has tons of memory, but the exception handling (and system call handling) is VERY CPU intensive. I already see this high CPU usage in my application that has a lot of interrupts and uses FPU in almost every thread.

Reply:
> I agree that with the current implementation FPU needs to be saved & restored always if the thread ever uses the FPU, even once. Otherwise the operation is never atomic and the restore occurs from the wrong stack frame. The current "lazy fpu" implementation is just plain wrong and needs to be removed completely in order for this to work.
> 
> However, I would still like to discuss doing the lazy FPU properly, by adding a separate FPU save area into the TCB. RISC-V has tons of memory, but the exception handling (and system call handling) is VERY CPU intensive. I already see this high CPU usage in my application that has a lot of interrupts and uses FPU in almost every thread.
> 

It's strange that almost every thread use FPU. Does the code use FPU instruction explicitly or generated by compler implicitely?

> I am in fact already writing the code, the CPU usage is just too important for my application.

It's fine to add a new option, but I am afraid the new option can't help so much as you expect if FPU instruction is inserted by compiler.

Reply:
Ok, let's wait your result.

Reply:
@xiaoxiang781216 

Finally got time to finalize the lazyFPU implementation and testing. Made a very simple code snippet with system calls enabled (to force trap entry):

```
int main(...)

... use floating point arithmetics ...

for (i = 0; i < n_samples; i++)
{
  t0 = read_cycles();
  pid = getpid();
  t1 = read_cycles();
  diff = t1 - t0;
  samples <- diff;
}

... print out avg, min, max
```

> Ok, let's wait your result.

With lazy FPU the system call takes:
avg:738, min:738, max:886

without lazy FPU + fix error in this PR  (save FPU if state not in {OFF, INIT} restore if state in {DIRTY, CLEAN}):
avg:954, min:954, max:1019

So on average it saves more than 200 cycles with this simple system call, ~22% (of course the benefit is less with more complex system calls).

Cumulative effect on my application that uses system calls + FPU is ~2.7%, so not insignificant at all.

Reply:
So, do you want to move the FPU to xcpt?

Reply:
Yes, I will provide patches for that.

Lazy FPU save/restore works in flat and protected mode but in kernel mode there is a problem. The user context is saved in user stack but there are corner cases where a different address environment is in use which cause a crash. binfmt_execmodule() is a good example, during that if a context switch occurs de-referencing tcb->xcp.regs will result in a crash due to reading from the wrong address environment. This will be fixed by moving the context save into the process's own kernel stack, I am working on this right now.

Reply:
could you add a option to keep the context on the stack since it could save 256bytes per no-FPU thread?

Reply:
You mean keep the current implementation as an option ? Because if lazy FPU is used we cannot store on the stack any more.

Of course it is doable, might need quite a bit of flagging though.